packages/sdk/src/protocols/filter/subscription_monitor.ts

```diff
@@ -0,0 +1,287 @@
+  ): boolean {
+    const hash = this.buildMessageHash(message);
+
+    this.verifiedPeers.add(peerId);
+    this.receivedMessagesFormPeer.add(`${peerId}-${hash}`);
+
+    if (this.receivedMessages.has(hash)) {
+      return true;
+    }
+
+    this.receivedMessages.add(hash);
+
+    return false;
+  }
+
+  private buildMessageHash(message: IProtoMessage): string {
+    return messageHashStr(this.pubsubTopic, message);
+  }
+
+  private startConnectionListener(): void {
+    this.connectionManager.addEventListener(
+      EConnectionStateEvents.CONNECTION_STATUS,
+      this.onConnectionChange as (v: CustomEvent<boolean>) => void
+    );
+  }
+
+  private stopConnectionListener(): void {
+    this.connectionManager.removeEventListener(
+      EConnectionStateEvents.CONNECTION_STATUS,
+      this.onConnectionChange as (v: CustomEvent<boolean>) => void
+    );
+  }
+
+  private async onConnectionChange({
+    detail: isConnected
+  }: CustomEvent<boolean>): Promise<void> {
+    if (!isConnected) {
+      this.stopKeepAlive();
+      return;
+    }
+
+    await Promise.all(this.peers.map((peer) => this.ping(peer, true)));
+    this.startKeepAlive();
+  }
+
+  private startKeepAlive(): void {
+    if (this.keepAliveIntervalId) {
+      return;
+    }
+
+    this.keepAliveIntervalId = setInterval(() => {
+      void this.peers.map((peer) => this.ping(peer));
+    }, this.config.keepAliveIntervalMs) as unknown as number;
+  }
+
+  private stopKeepAlive(): void {
+    if (!this.keepAliveIntervalId) {
+      return;
+    }
+
+    clearInterval(this.keepAliveIntervalId);
+    this.keepAliveIntervalId = undefined;
+  }
+
+  private startPeerConnectionListener(): void {
+    this.libp2p.addEventListener(
+      "peer:connect",
+      this.onPeerConnected as EventHandler<CustomEvent<PeerId | undefined>>
+    );
+    this.libp2p.addEventListener(
+      "peer:disconnect",
+      this.onPeerDisconnected as EventHandler<CustomEvent<PeerId | undefined>>
+    );
+  }
+
+  private stopPeerConnectionListener(): void {
+    this.libp2p.removeEventListener(
+      "peer:connect",
+      this.onPeerConnected as EventHandler<CustomEvent<PeerId | undefined>>
+    );
+    this.libp2p.removeEventListener(
+      "peer:disconnect",
+      this.onPeerDisconnected as EventHandler<CustomEvent<PeerId | undefined>>
+    );
+  }
+
+  private async onPeerConnected(_event: CustomEvent<PeerId>): Promise<void> {
+    // TODO(weboko): use config.numOfUsedPeers here
+    if (this.peers.length > 0) {
+      return;
+    }
+
+    this.peers = await this.peerManager.getPeers();
+    await Promise.all(this.peers.map((peer) => this.subscribe(peer)));
+  }
+
+  private async onPeerDisconnected(event: CustomEvent<PeerId>): Promise<void> {
+    const hasNotBeenUsed = !this.peers.find((p) => p.id.equals(event.detail));
+    if (hasNotBeenUsed) {
+      return;
+    }
+
+    this.peers = await this.peerManager.getPeers();
+    await Promise.all(this.peers.map((peer) => this.subscribe(peer)));
```

Comment:
similar to above, a comment would be helpful. it appears that if a disconnected peer had been used to establish a filter subscription, then this triggers an attempt to resubscribe to all peers?

Reply:
precisely, added a comment